pkg/policy/repository.go

```diff
@@ -226,24 +261,23 @@ func (p *Repository) allowsL4Ingress(ctx *SearchContext) api.Decision {
 // held.
 func (p *Repository) AllowsRLocked(ctx *SearchContext) api.Decision {
 	ctx.PolicyTrace("Tracing %s\n", ctx.String())
-	decision := p.CanReachRLocked(ctx)
+	decision := p.CanReachRLocked(*ctx)
 	ctx.PolicyTrace("Label verdict: %s", decision.String())
 	if decision == api.Allowed {
 		return decision
 	}
 
-	// We only report the overall decision as L4 inclusive if a port has
-	// been specified
-	if len(ctx.DPorts) != 0 {
```

Comment:
That's because the resolution of the ingress visibility rules synthesize new L3-only rules in some cases. More specifically, in case the policy enforcement mode is "default", and there are no Ingress or Egress rules matching, but there are one or more ingress visibility rules matching.
In this case, the `L4Policy` will contain one "allow-all" Ingress rule and one "allow-all" Egress rule created by the visibility resolution. Those rules are synthesized, so they can't be evaluated just by iterating on the `Repository.rules`. That requires resolving the `L4Policy`.

That's another argument for designing a new "Policy" type that encompasses all of entities, CIDR, L3, L4, visibility, etc. and that should be used for all decisions and tracing.

Reply:
OK, thanks for the explanation.